internal/graph/lookupresources.go

```diff
@@ -44,105 +45,113 @@ func (cl *CursoredLookupResources) LookupResources(
+		reachableResourcesFound := 0
+
+		// Create a new handling stream that consumes the reachable resources results and publishes them
+		// as found resources if they are properly checked.
+		rrStream := dispatch.NewHandlingDispatchStream(ctx, func(result *v1.DispatchReachableResourcesResponse) error {
+			reachableResourcesCursor = result.AfterResponseCursor
+			reachableResourcesFound++
+
+			if limits.hasExhaustedLimit() {
+				return nil
 			}
-			checkResults = results
-			checkMeta = addCallToResponseMetadata(resultsMeta)
-		}
 
-		for _, reachableResource := range result.Resources {
-			switch reachableResource.ResultStatus {
-			case v1.ReachableResource_HAS_PERMISSION:
-				stream.Publish(&v1.DispatchLookupResourcesResponse{
-					ResolvedResources: []*v1.ResolvedResource{
-						{
-							ResourceId:     reachableResource.ResourceId,
-							Permissionship: v1.ResolvedResource_HAS_PERMISSION,
-						},
+			reachableResource := result.Resource
+			metadata := addCallToResponseMetadata(result.Metadata)
+
+			// See if a check is required.
+			var checkResult *v1.ResourceCheckResult
+			if reachableResource.ResultStatus == v1.ReachableResource_REQUIRES_CHECK {
+				results, resultsMeta, err := computed.ComputeBulkCheck(
+					ctx,
+					cl.c,
+					computed.CheckParameters{
+						ResourceType:  req.ObjectRelation,
+						Subject:       req.Subject,
+						CaveatContext: req.Context.AsMap(),
+						AtRevision:    req.Revision,
+						MaximumDepth:  req.Metadata.DepthRemaining,
+						DebugOption:   computed.NoDebugging,
 					},
-					Metadata: addCallToResponseMetadata(result.Metadata),
-				})
-				continue
-
-			case v1.ReachableResource_REQUIRES_CHECK:
-				checkResult, ok := checkResults[reachableResource.ResourceId]
-				if !ok {
-					return spiceerrors.MustBugf("missing checked result")
-				}
-				switch checkResult.Membership {
-				case v1.ResourceCheckResult_MEMBER:
-					stream.Publish(&v1.DispatchLookupResourcesResponse{
-						ResolvedResources: []*v1.ResolvedResource{
-							{
-								ResourceId:     reachableResource.ResourceId,
-								Permissionship: v1.ResolvedResource_HAS_PERMISSION,
-							},
-						},
-						Metadata: combineResponseMetadata(result.Metadata, checkMeta),
-					})
-					continue
-
-				case v1.ResourceCheckResult_CAVEATED_MEMBER:
-					stream.Publish(&v1.DispatchLookupResourcesResponse{
-						ResolvedResources: []*v1.ResolvedResource{
-							{
-								ResourceId:             reachableResource.ResourceId,
-								Permissionship:         v1.ResolvedResource_CONDITIONALLY_HAS_PERMISSION,
-								MissingRequiredContext: checkResult.MissingExprFields,
-							},
-						},
-						Metadata: combineResponseMetadata(result.Metadata, checkMeta),
-					})
-					continue
-
-				case v1.ResourceCheckResult_NOT_MEMBER:
-					// Skip.
-					continue
-
-				default:
-					return spiceerrors.MustBugf("unknown check result status for reachable resources")
+					[]string{reachableResource.ResourceId},
```

Comment:
Reachable resources is collecting results "in the background", if on another node.

It would be better to buffer X elements, and I intend to do that as a followup, but choosing the correct count has its own issues and I didn't want to make this PR even larger

Reply:
Let's open a follow up issue please 🙏🏻